### PR TITLE
[flow] Add ops and passes for dynamicize static shapes

### PR DESCRIPTION
**Pull request by antiagainst**
This commit defines two operations, `dispatch.dynamicize_shape`,
and `dispatch.dynamicize_dim`, and a pass for turning static
shapes into dynamic ones around dispatch region boundary.

Dynamicizing input shapes enables better executable deduplication
at the flow level, so that kernel CodeGen flow is not forced to
specialize towards every input shapes. Instead, kernel CodeGen
backends will see dynamically shaped executables, with all
concrete static dimension values as `stream.values` attributes,
and make local decisions w.r.t. whether/what to specialize or
create multiple versions.

Special ops are needed for this conversion, to act as barrier for
optimizations across region boundaries and hold up ABI consistency.

Progress towards https://github.com/openxla/iree/issues/12230

### CODE REVIEW

**Comment by benvanik** [comment]
(will take a look at this today!)

**Review by benvanik — CHANGES_REQUESTED** [review]
Sorry for the delay; still processing (flu and meetings x_x), but there's some interactions I've had to noodle on.

An easy one is that these ops shouldn't survive all the way to the stream dialect but instead should be removed as part of dispatch workgroups creation (dynamicize_dim -> arith.constant, dynamicize_shape -> flow.tensor.reshape) or immediately after in a cleanup pass - I think CaptureDispatchDynamicDimsPass is probably the best place to do the cleanup (after that point we don't care, it's run right after formation, and it pairs well semantically with what this pass does). We don't want these leaking out past where they are needed (just around dispatch region formation).

Another is that I'm sketched out by the empty/insert/extract sniffing happening in the pass - we want to be able to support any tensor ops including ones we don't own (custom ops from users, new dialects, etc). Are there interfaces that could be used instead? We can't grow a table for all ops in all dialects that may need updates. Right now this is risky for compilation failures and execution correctness as if you don't have those special cases it'd be creating invalid ops that would immediately fail verification (hopefully) or cause chaos - we don't want to play that whackamole with ops that may come through that aren't special cased. I think we could use this for experimentation but we'd need to solve this if we want to turn this on (and I'd really like to :). There are things like trying the generic conversion and running the per-op verifier but that requires that all ops implement verifiers correctly (if even possible, such as on external function calls/custom ops).

Naming-wise `flow.dispatch.` is used for ops that are dispatches or are inside of dispatches and `flow.tensor.` is used for tensor operations outside of dispatches like these so we'll want to make these consistent with that scheme. Related, the dynamicize_shape op should implement the tied op interface like `flow.tensor.reshape` as otherwise it breaks semantics (the dynamicized tensor is a distinct instance from its input today, which breaks things like out params). I'm wondering if we can't just use `flow.tensor.reshape` instead of having a new op - it has folders and such but if the dimension values feeding in are the special dynamicized ones then it should be fine.

If we can reuse `flow.tensor.reshape` then that just leaves dynamicize_dim, which is effectively just a `util.unfoldable_constant` but with a unique ID so that we can easily clean it up. I think we could change dynamicize_dim to `flow.dynamic_constant` as it's just an optimization barrier and not related to dims specifically. When used for purposes like this it's actually a good frontend op that may get inserted early in the pipeline when the user knows they want something to remain dynamic (a dim or any other parameter).

Summary:
- remove these ops in CaptureDispatchDynamicDimsPass so they don't live any longer than required
- make more robust/plan for how we can land this enabled without a giant op switch in the pass (we could land the PR disabled, but I want to know this isn't going to never be able to graduate)
- see if flow.tensor.reshape works instead of dynamicize shape
- rename the ops for consistency

happy to gvc too (having trouble forming thoughts so let me know if this makes sense :)

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DynamicizeStaticShapes.cpp [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+/// Returns a vector of SSA values representing each dimension of the given
+/// `value`'s shape.
+SmallVector<Value, 4> getDimValues(OpBuilder &builder, Location loc,
+                                   Value value, ValueRange dynamicDims) {
+  auto type = value.getType().cast<ShapedType>();
+  SmallVector<Value, 4> dimValues;
+  dimValues.reserve(type.getRank());
+
+  int dimIndex = 0;
+  for (int64_t dimSize : type.getShape()) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/DynamicizeStaticShapes.cpp:68]: You can use https://github.com/llvm/llvm-project/blob/1def3141135c072a1d3e51e82e113dd67b0def97/mlir/include/mlir/Dialect/Tensor/Utils/Utils.h#LL36C55-L36C55 instead.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DynamicizeStaticShapes.cpp [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+/// `partialType` from a `fullType`, updates `offsetValues` and `sizeValues`
+/// with SSA values for the offsets and sizes.
+void getTensorSliceOffsetsSizesAndDimValues(
+    OpBuilder &builder, Location loc, OffsetSizeAndStrideOpInterface op,
+    RankedTensorType fullType, RankedTensorType partialType,
+    SmallVectorImpl<OpFoldResult> &offsetValues,
+    SmallVectorImpl<OpFoldResult> &sizeValues) {
+  const SmallVector<OpFoldResult> offsets = op.getMixedOffsets();
+  offsetValues.resize(offsets.size());
+  for (int i = 0, e = offsets.size(); i < e; ++i) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/DynamicizeStaticShapes.cpp:90]: Nit: This is a bit mouthful.

```
auto offsets = llvm::map_range(op.getMixedOffsets(), [&](OpFoldResult ofr) {
  
}
```

seems easier to read.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DynamicizeStaticShapes.cpp [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    value.replaceUsesWithIf(newValue, [regionOp](OpOperand &operand) {
+      return operand.getOwner()->getParentOp() == regionOp;
+    });
+  }
+
+  // Then update all compute ops in the dispatch region to return values of
+  // dynamic shapes.
+  Block &block = regionOp.getBody().getBlocks().front();
+  SmallVector<Operation *> tensorOps;
+  for (Operation &op : block.without_terminator()) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/DynamicizeStaticShapes.cpp:157]: I think we need a worklist here to transitively follow the uses..... similar to https://github.com/openxla/iree/blob/62ee2f342484a1468abe5bc96a2be7c3192a6e63/compiler/src/iree/compiler/Codegen/Utils/Utils.h#L172

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

**Comment by antiagainst** [comment]
@benvanik: thanks for the reviews! :) Yeah this is meant to be an early draft to seek feedback. The points you made in the above generally makes sense to me. I've some more thoughts, so let's sync up in VC later.

**Comment by benvanik** [comment]
Closing stale PR.

**Pull request closed without merging**